### PR TITLE
Fix macro imports

### PR DESCRIPTION
**Pull request by brainkim**
### What does this PR do?

Currently, macros cannot be imported using a namespace import, default imports, or with named import aliases.

```js
// All of the following imports will fail:
import defaultMacro from "macros.js" with {type: "macro"};
import * as macros from "macros.js" with {type: "macro"};
import {macro as myMacro} from "macros.js" with {type: "macro"};
```

Fixes https://github.com/oven-sh/bun/issues/15348

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

I wrote automated tests

- [x] I included a test for the new code, or existing tests cover it
- [x] I ran my tests locally and they pass (`bun-debug test test-file-name.test`)
- [x] I checked the lifetime of memory allocated to verify it's (1) freed and (2) only freed when it should be
- [x] I included a test for the new code, or an existing test covers it
- [x] JSValue used outside outside of the stack is either wrapped in a JSC.Strong or is JSValueProtect'ed
- [x] I wrote TypeScript/JavaScript tests and they pass locally (`bun-debug test test-file-name.test`)

## TODO:

- [ ] increment expected_version in RuntimeTranspilerCache?
- [x] test situations with export aliases or re-exports

### CODE REVIEW

File: src/js_parser.zig [hunk 1]
```diff
@@ -16499,12 +16522,15 @@ fn NewParser_(
                         e_.tag = p.visitExpr(tag);
 
                         if (comptime allow_macros) {
```
Comment (brainkim) [inline src/js_parser.zig]: I updated the code for macro template tags but it doesn’t seem like Bun supports template tag macros yet...

File: src/js_parser.zig [hunk 2]
```diff
@@ -17425,8 +17460,13 @@ fn NewParser_(
 
                     if (comptime allow_macros) {
                         if (is_macro_ref and !p.options.features.is_macro_runtime) {
-                            const ref = e_.target.data.e_import_identifier.ref;
-                            const import_record_id = p.macro.refs.get(ref).?;
+                            const ref = switch (e_.target.data) {
+                                .e_import_identifier => |ident| ident.ref,
+                                .e_dot => |dot| dot.target.data.e_identifier.ref,
```
Comment (brainkim) [inline src/js_parser.zig]: I’m not really sure why `.e_dot` expressions have a target which is `.e_identifier` and not `.e_import_identifier` when the target is a namespace identifier, but I’m too afraid to change it.

File: src/js_parser.zig [hunk 3]
```diff
@@ -8969,6 +8990,8 @@ fn NewParser_(
                         .import_record_index = stmt.import_record_index,
                     }) catch unreachable;
                 }
+
+                // TODO: not sure how to handle macro remappings for namespace imports
```
Comment (brainkim) [inline src/js_parser.zig:8994]: I’m not too sure about how the package.json macro remapping feature works. I tried to do what I can, but to allow namepspace imports to be remapped would probably be pretty hard as far as I can tell.

**Comment by Jarred-Sumner** [comment]
I think this approach overall makes sense.

File: src/js_parser.zig (hunk 2 above)
Comment (Jarred-Sumner) [reply src/js_parser.zig]: We remove these imports before they reach the final output

File: test/bundler/transpiler/macro-test.test.ts [hunk 4]
```diff
@@ -1,6 +1,12 @@
 import { escapeHTML } from "bun" assert { type: "macro" };
 import { expect, test } from "bun:test";
 import { addStrings, addStringsUTF16, escape, identity } from "./macro.ts" assert { type: "macro" };
+import defaultMacro, {
+  identity as identity1,
+  identity as identity2,
+} from "./macro.ts" assert { type: "macro" };
+
+import * as macros from "./macro.ts" assert { type: "macro" };
```
Comment (Jarred-Sumner) [inline test/bundler/transpiler/macro-test.test.ts:10]: Can you add one more test that aliases a default import?

```js
import { default as defaultMacro2 } from './macro.ts" assert {type: "macro"};
```

**Review by Jarred-Sumner — APPROVED** [review]
Looks good assuming the relevant tests pass and it doesn't cause any regressions (I don't think it will cause any regressions) 

The main thing I'd be curious about is how we generate the wrapper function in the macro to register the callback to call. I haven't looked at that code in awhile.

**Comment by brainkim** [comment]
@Jarred-Sumner The macros are generated using an entrypoint in src/bundler/entry_points.zig

The exports are imported with a top-level await on an `import` expression.

```zig
// This is not very fast. The idea is: we want to generate a unique entry point
// per macro function export that registers the macro Registering the macro
// happens in VirtualMachine We "register" it which just marks the JSValue as
// protected. This is mostly a workaround for being unable to call ESM exported
// functions from C++. When that is resolved, we should remove this.
pub const MacroEntryPoint = struct {
    code_buffer: [bun.MAX_PATH_BYTES * 2 + 500]u8 = undefined,
    output_code_buffer: [bun.MAX_PATH_BYTES * 8 + 500]u8 = undefined,
    source: logger.Source = undefined,

    pub fn generateID(entry_path: string, function_name: string, buf: []u8, len: *u32) i32 {
        var hasher = bun.Wyhash11.init(0);
        hasher.update(js_ast.Macro.namespaceWithColon);
        hasher.update(entry_path);
        hasher.update(function_name);
        const hash = hasher.final();
        const fmt = bun.fmt.hexIntLower(hash);

        const specifier = std.fmt.bufPrint(buf, js_ast.Macro.namespaceWithColon ++ "//{any}.js", .{fmt}) catch unreachable;
        len.* = @as(u32, @truncate(specifier.len));

        return generateIDFromSpecifier(specifier);
    }

    pub fn generateIDFromSpecifier(specifier: string) i32 {
        return @as(i32, @bitCast(@as(u32, @truncate(bun.hash(specifier)))));
    }

    pub fn generate(
        entry: *MacroEntryPoint,
        _: *Bundler,
        import_path: Fs.PathName,
        function_name: string,
        macro_id: i32,
        macro_label_: string,
    ) !void {
        const dir_to_use: string = if (import_path.dir.len == 0) "" else import_path.dirWithTrailingSlash();
        bun.copy(u8, &entry.code_buffer, macro_label_);
        const macro_label = entry.code_buffer[0..macro_label_.len];

        const code = brk: {
            if (strings.eqlComptime(import_path.base, "bun")) {
                break :brk try std.fmt.bufPrint(
                    entry.code_buffer[macro_label.len..],
                    \\//Auto-generated file
                    \\var Macros;
                    \\try {{
                    \\  Macros = globalThis.Bun;
                    \\}} catch (err) {{
                    \\   console.error("Error importing macro");
                    \\   throw err;
                    \\}}
                    \\const macro = Macros['{s}'];
                    \\if (!macro) {{
                    \\  throw new Error("Macro '{s}' not found in 'bun'");
                    \\}}
                    \\
                    \\Bun.registerMacro({d}, macro);
                ,
                    .{
                        function_name,
                        function_name,
                        macro_id,
                    },
                );
            }

            break :brk try std.fmt.bufPrint(
                entry.code_buffer[macro_label.len..],
                \\//Auto-generated file
                \\var Macros;
                \\try {{
                \\  Macros = await import('{s}{s}');
                \\}} catch (err) {{
                \\   console.error("Error importing macro");
                \\   throw err;
                \\}}
                \\if (!('{s}' in Macros)) {{
                \\  throw new Error("Macro '{s}' not found in '{s}{s}'");
                \\}}
                \\
                \\Bun.registerMacro({d}, Macros['{s}']);
            ,
                .{
                    bun.fmt.fmtPath(u8, dir_to_use, .{
                        .escape_backslashes = true,
                    }),
                    bun.fmt.fmtPath(u8, import_path.filename, .{
                        .escape_backslashes = true,
                    }),
                    function_name,
                    function_name,
                    bun.fmt.fmtPath(u8, dir_to_use, .{
                        .escape_backslashes = true,
                    }),
                    bun.fmt.fmtPath(u8, import_path.filename, .{
                        .escape_backslashes = true,
                    }),
                    macro_id,
                    function_name,
                },
            );
        };

        entry.source = logger.Source.initPathString(macro_label, code);
        entry.source.path.text = macro_label;
        entry.source.path.namespace = js_ast.Macro.namespace;
    }
};
```

I thought I might have to do additional work to handle re-exports, but this seems to be robust so long as you have the name of the export.

**Review by paperclover — APPROVED** [review]

**Pull request merged**